File: lib/core/log_service.dart
import 'dart:collection';
import 'package:flutter/foundation.dart';

enum LogLevel { debug, info, warn, error }

class LogEntry {
  final DateTime timestamp;
  final LogLevel level;
  final String message;
  final String source;

  LogEntry({
    required this.level,
    required this.message,
    this.source = '',
    DateTime? timestamp,
  }) : timestamp = timestamp ?? DateTime.now();

  String get formattedTime {
    final h = timestamp.hour.toString().padLeft(2, '0');
    final m = timestamp.minute.toString().padLeft(2, '0');
    final s = timestamp.second.toString().padLeft(2, '0');
    final ms = timestamp.millisecond.toString().padLeft(3, '0');
    return '$h:$m:$s.$ms';
  }

  String get levelLabel {
    switch (level) {
      case LogLevel.debug:
        return 'DEBUG';
      case LogLevel.info:
        return 'INFO';
      case LogLevel.warn:
        return 'WARN';
      case LogLevel.error:
        return 'ERROR';
    }
  }

  String toFormattedString() {
    final src = source.isNotEmpty ? ' [$source]' : '';
    return '[$formattedTime] [$levelLabel]$src $message';
  }
}

class LogService extends ChangeNotifier {
  static final LogService _instance = LogService._();
  static LogService get instance => _instance;

  final List<LogEntry> _logs = [];
  static const int _maxLogs = 2000;
  LogLevel _minimumLevel = LogLevel.debug;

  LogLevel get minimumLevel => _minimumLevel;
  set minimumLevel(LogLevel level) {
    _minimumLevel = level;
    notifyListeners();
  }

  LogService._();

  UnmodifiableListView<LogEntry> get logs => UnmodifiableListView(_logs);

  void debug(String message, {String source = ''}) {
    _addLog(LogLevel.debug, message, source: source);
  }

  void info(String message, {String source = ''}) {
    _addLog(LogLevel.info, message, source: source);
  }

  void warn(String message, {String source = ''}) {
    _addLog(LogLevel.warn, message, source: source);
  }

  void error(String message, {String source = ''}) {
    _addLog(LogLevel.error, message, source: source);
  }

  void _addLog(LogLevel level, String message, {String source = ''}) {
    if (level.index < _minimumLevel.index) return;
    final entry = LogEntry(level: level, message: message, source: source);
    _logs.add(entry);
    if (_logs.length > _maxLogs) {
      _logs.removeAt(0);
    }
    notifyListeners();
  }

  void clear() {
    _logs.clear();
    notifyListeners();
  }

  String getPlainText() {
    return _logs.map((e) => e.toFormattedString()).join('\n');
  }

  List<LogEntry> getFilteredLogs({LogLevel? minLevel, String? query}) {
    var result = _logs.where((e) => minLevel == null || e.level.index >= minLevel.index);
    if (query != null && query.isNotEmpty) {
      final q = query.toLowerCase();
      result = result.where((e) =>
          e.message.toLowerCase().contains(q) ||
          e.source.toLowerCase().contains(q));
    }
    return result.toList();
  }
}

File: android/app/src/main/java/com/easytier/service/LogService.kt
package com.easytier.service

import com.easytier.data.LogEntry
import com.easytier.data.LogLevel

/** 内存日志服务，带级别过滤和上限 */
object LogService {
    private val _logs = mutableListOf<LogEntry>()
    private const val MAX_LOGS = 2000
    var minimumLevel: LogLevel = LogLevel.DEBUG

    val logs: List<LogEntry> get() = _logs.toList()

    fun debug(message: String, source: String = "") = addLog(LogLevel.DEBUG, message, source)
    fun info(message: String, source: String = "") = addLog(LogLevel.INFO, message, source)
    fun warn(message: String, source: String = "") = addLog(LogLevel.WARN, message, source)
    fun error(message: String, source: String = "") = addLog(LogLevel.ERROR, message, source)

    private fun addLog(level: LogLevel, message: String, source: String) {
        if (level.ordinal < minimumLevel.ordinal) return
        val entry = LogEntry(level = level, message = message, source = source)
        synchronized(_logs) {
            _logs.add(entry)
            if (_logs.size > MAX_LOGS) _logs.removeAt(0)
        }
    }

    fun clear() { synchronized(_logs) { _logs.clear() } }

    fun getPlainText(): String = synchronized(_logs) {
        _logs.joinToString("\n") { it.toFormattedString() }
    }

    fun getFilteredLogs(minLevel: LogLevel? = null, query: String? = null): List<LogEntry> =
        synchronized(_logs) {
            var result: List<LogEntry> = _logs
            if (minLevel != null) result = result.filter { it.level.ordinal >= minLevel.ordinal }
            if (!query.isNullOrEmpty()) {
                val q = query.lowercase()
                result = result.filter {
                    it.message.lowercase().contains(q) || it.source.lowercase().contains(q)
                }
            }
            result
        }
}
